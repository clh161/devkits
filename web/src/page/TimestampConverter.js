// @flow strict
import type { Node } from 'react';
import React, { useState } from 'react';
import { Grid, TextField, Typography } from '@material-ui/core';
import { Helmet } from 'react-helmet';

type DatetimeConfig = {
  name: string,
  value: number,
};

type Props = {
  initTimestamp?: number,
};

export default function TimestampConverter({ initTimestamp }: Props): Node {
  const [unixTimestamp, setUnixTimestamp] = useState(
    parseInt(initTimestamp ?? Date.now() / 1000)
  );
  const [datetime, setDatetime] = useState(unixToDate(parseInt(unixTimestamp)));

  function onUnixTimestampChange(event) {
    const { value } = event.target;
    setUnixTimestamp(value);
    setDatetime(unixToDate(Number(value)));
  }

  function onDatetimeChange(name, event) {
    const { value } = event.target;
    setDatetime(
      datetime.map((dt) => {
        if (name === dt.name) {
          dt.value = value;
        }
        return dt;
      })
    );
    const datestamp =
      new Date(
        Date.UTC(
          ...datetime.map((dt) => {
            if (dt.name === 'Month') {
              return dt.value - 1;
            }
            return dt.value;
          })
        )
      ).getTime() / 1000;
    setUnixTimestamp(datestamp);
  }

  return (
    <div>
      <Helmet>
        <title>Unix Timestamp Convertor (UTC)</title>
        <meta
          content="unix timestamp and date time converter"
          name="description"
        />
        <meta content="timestamp, epoch, converter" name="keywords" />
      </Helmet>

      <Grid container spacing={3}>
        <Grid item xs={12}>
          <Typography component="h1" variant="h4">
            Unix Timestamp Convertor (UTC)
          </Typography>
        </Grid>
        <Grid item xs={12}>
          <TextField
            label="Timestamp"
            onChange={onUnixTimestampChange}
            type="number"
            value={unixTimestamp}
          />
        </Grid>
        {datetime.map((dt: DatetimeConfig) => {
          return (
            <Grid item key={dt.name} xs={2}>
              <TextField
                label={dt.name}
                onChange={(event) => {
                  onDatetimeChange(dt.name, event);
                }}
                type="number"
                value={dt.value}
              />
            </Grid>
          );
        })}
      </Grid>
    </div>
  );
}

export function unixToDate(timestamp: number): Array<DatetimeConfig> {
  const date = new Date(timestamp * 1000);

  return [
    {
      name: 'Year',
      value: date.getUTCFullYear(),
    },
    {
      name: 'Month',
      value: date.getUTCMonth() + 1,
    },
    {
      name: 'Day',
      value: date.getDate(),
    },
    {
      name: 'Hour',
      value: date.getUTCHours(),
    },
    {
      name: 'Minute',
      value: date.getUTCMinutes(),
    },
    {
      name: 'Second',
      value: date.getUTCSeconds(),
    },
  ];
}
