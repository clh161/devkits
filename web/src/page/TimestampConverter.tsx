import { Grid, TextField } from '@material-ui/core';
import React, { ReactElement, useState } from 'react';
type DatetimeConfig = {
  name: string;
  value: number;
};
type Props = {
  initTimestamp?: number;
};
export default function TimestampConverter({
  initTimestamp,
}: Props): ReactElement {
  const [unixTimestamp, setUnixTimestamp] = useState(
    parseInt(initTimestamp?.toString() ?? (Date.now() / 1000).toString())
  );
  const [datetime, setDatetime] = useState(
    unixToDate(parseInt(unixTimestamp.toString()))
  );

  function onUnixTimestampChange(event) {
    const { value } = event.target;
    setUnixTimestamp(value);
    setDatetime(unixToDate(Number(value)));
  }

  function onDatetimeChange(name, event) {
    const { value } = event.target;
    setDatetime(
      datetime.map((dt) => {
        if (name === dt.name) {
          dt.value = value;
        }

        return dt;
      })
    );
    const time = datetime.map((dt) => {
      if (dt.name === 'Month') {
        return dt.value - 1;
      }

      return dt.value;
    });

    const datestamp =
      new Date(
        Date.UTC(time[0], time[1], time[2], time[3], time[4], time[5])
      ).getTime() / 1000;
    setUnixTimestamp(datestamp);
  }

  return (
    <Grid container spacing={3}>
      <Grid item xs={12}>
        <TextField
          label='Timestamp'
          onChange={onUnixTimestampChange}
          type='number'
          value={unixTimestamp}
        />
      </Grid>
      {datetime.map((dt: DatetimeConfig) => {
        return (
          <Grid item key={dt.name} xs={2}>
            <TextField
              label={dt.name}
              onChange={(event) => {
                onDatetimeChange(dt.name, event);
              }}
              type='number'
              value={dt.value}
            />
          </Grid>
        );
      })}
    </Grid>
  );
}
export function unixToDate(timestamp: number): Array<DatetimeConfig> {
  const date = new Date(timestamp * 1000);
  return [
    {
      name: 'Year',
      value: date.getUTCFullYear(),
    },
    {
      name: 'Month',
      value: date.getUTCMonth() + 1,
    },
    {
      name: 'Day',
      value: date.getDate(),
    },
    {
      name: 'Hour',
      value: date.getUTCHours(),
    },
    {
      name: 'Minute',
      value: date.getUTCMinutes(),
    },
    {
      name: 'Second',
      value: date.getUTCSeconds(),
    },
  ];
}
