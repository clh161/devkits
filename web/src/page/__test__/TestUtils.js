// @flow strict

export function mockLocation(mockPath: string) {
  jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: () => ({
      pathname: mockPath,
    }),
  }));
}
