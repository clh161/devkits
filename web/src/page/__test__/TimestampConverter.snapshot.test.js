// @flow

import React from 'react';
import { Helmet } from 'react-helmet';
import renderer from 'react-test-renderer';

import TimestampConverter from '../TimestampConverter';
import { mockLocation } from './TestUtils';

mockLocation('/unix-timestamp-converter');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '/unix-timestamp-converter',
  }),
}));
it('renders correctly', () => {
  const tree = renderer
    .create(<TimestampConverter initTimestamp={1620432665} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
  expect(Helmet.peek()).toMatchSnapshot();
});
