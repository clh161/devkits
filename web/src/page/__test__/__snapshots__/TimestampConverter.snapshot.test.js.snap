// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h1
        className="MuiTypography-root MuiTypography-h4"
      >
        Unix Timestamp Convertor (UTC)
      </h1>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Timestamp
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={1620432665}
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Year
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={2021}
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Month
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={5}
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Day
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={8}
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Hour
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={0}
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Minute
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={11}
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Second
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="number"
            value={5}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
